### PR TITLE
Suggestions page

### PR DESCRIPTION
**Pull request by tomrf1**
1. gets most-read 20 articles from ophan
2. uses the tags to get suggested snippet atoms from Targeting api
3. for each article which does not already have a snippet embedded, displays a list of suggested atoms

picture 16

### CODE REVIEW

File: public/js/actions/AtomActions/getSuggestionsForLatestContent.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        const urls = filtered.map(target => target.url);
+        if (filtered.length > 0) {
+          tagToTargetAtoms[tag] = (tagToTargetAtoms[tag]) ? tagToTargetAtoms[tag].tag(urls) : urls;
+        }
+        return Promise.resolve();
+      });
+  };
+
+  //Sequentially fetch targets for each tag
+  return tags.reduce((p, tag) => p.then(() => fetch(tag)), Promise.resolve())
```
Comment (regiskuckaertz) [inline public/js/actions/AtomActions/getSuggestionsForLatestContent.js]: You can thread `tapToTargetAtoms` in the promise chain if you prefer that style:
```
const fetch = (tag, tapToTargetAtoms) => fetchTargetsForTag(tag)
  .then(targets =>
    ({ ...tapToTargetAtoms,
      [tag]: target.filter(...).map(...)
    }));

return tags.reduce((p, tag) => p.then(ttta => fetch(tag, ttta)), Promise.resolve({});
```

File: public/js/actions/AtomActions/getSuggestionsForLatestContent.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    .then(() => {
+      return Promise.resolve(tagToTargetAtoms);
+    });
+}
+
+//Deduplicate the targeting urls and retrieve the atoms
+function getAtomUrlToAtom(tagToUrls) {
+  const atomUrlToAtom = {};
+
+  const atomUrls = distinct(flatten(Object.keys(tagToUrls).map(tag => tagToUrls[tag])));
```
Comment (regiskuckaertz) [inline public/js/actions/AtomActions/getSuggestionsForLatestContent.js]: `const atomUrls = distinct(flatten(Object.values(tagToUrls)))` should work

File: public/js/actions/AtomActions/getSuggestionsForLatestContent.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  //Sequentially fetch targets for each tag
+  return tags.reduce((p, tag) => p.then(() => fetch(tag)), Promise.resolve())
+    .then(() => {
+      return Promise.resolve(tagToTargetAtoms);
+    });
+}
+
+//Deduplicate the targeting urls and retrieve the atoms
+function getAtomUrlToAtom(tagToUrls) {
+  const atomUrlToAtom = {};
```
Comment (regiskuckaertz) [inline public/js/actions/AtomActions/getSuggestionsForLatestContent.js]: You can move that var into the last `.then` callback

File: public/js/services/TargetingApi.js [hunk 4]
```diff
@@ -62,3 +62,37 @@ export const deleteTarget = (targetId) => {
...
   );
 };
+
+export const fetchTargetsForTags = (tags) => {
+
+  const store = getStore();
+  const state = store.getState();
+  const targetingUrl = state.config.targetingUrl;
+
+  const tagsParams = tags.map(tag => `tags=${tag}`).join('&');
```
Comment (regiskuckaertz) [inline public/js/services/TargetingApi.js]: `tags=${encodeURIComponent(tag)}`?

File: public/js/actions/AtomActions/getSuggestionsForLatestContent.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+function getTagToTargetUrls(tags) {
+  const tagToTargetAtoms = {};
+
+  const fetch = (tag) => {
+    return fetchTargetsForTag(tag)
+      .then(targets => {
+        const filtered = targets.filter(target => target.url.includes("/atom/"));
+        const urls = filtered.map(target => target.url);
+        if (filtered.length > 0) {
+          tagToTargetAtoms[tag] = (tagToTargetAtoms[tag]) ? tagToTargetAtoms[tag].tag(urls) : urls;
```
Comment (regiskuckaertz) [inline public/js/actions/AtomActions/getSuggestionsForLatestContent.js]: Regardless of which style you prefer, I don't think the consequent is every going to be called here. As each tag is unique, you can just write `tagToTargetAtoms[tag] = urls`.

**Comment by regiskuckaertz** [comment]
Solid stuff! It all looks good to me, except for that last comment with urls 👍

File: public/js/actions/AtomActions/getSuggestionsForLatestContent.js (hunk 5 above)
Comment (tomrf1) [reply public/js/actions/AtomActions/getSuggestionsForLatestContent.js]: good point, I'm really not sure why I wrote that...

**Pull request merged**